server/src/main/java/org/elasticsearch/common/io/stream/NetworkStreamOutput.java

```diff
@@ -0,0 +1,227 @@
+    private int currentCapacity = 0;
+    private int currentPageOffset;
+
+    public NetworkStreamOutput() {
+        this(new BytesRefRecycler(PageCacheRecycler.NON_RECYCLING_INSTANCE));
+    }
+
+    public NetworkStreamOutput(Recycler<BytesRef> recycler) {
+        this.recycler = recycler;
+        try (Recycler.V<BytesRef> obtain = recycler.obtain()) {
+            pageSize = obtain.v().length;
+        }
+        this.currentPageOffset = pageSize;
+    }
+
+    @Override
+    public long position() {
+        return ((long) pageSize * pageIndex) + currentPageOffset;
+    }
+
+    @Override
+    public void writeByte(byte b) {
+        ensureCapacity(1);
+        BytesRef currentPage = pages.get(pageIndex).v();
+        currentPage.bytes[currentPage.offset + currentPageOffset] = b;
+        currentPageOffset++;
+    }
+
+    @Override
+    public void writeBytes(byte[] b, int offset, int length) {
+        // nothing to copy
+        if (length == 0) {
+            return;
+        }
+
+        // illegal args: offset and/or length exceed array size
+        if (b.length < (offset + length)) {
+            throw new IllegalArgumentException("Illegal offset " + offset + "/length " + length + " for byte[] of length " + b.length);
+        }
+
+        // get enough pages for new size
+        ensureCapacity(length);
+
+        // bulk copy
+        int bytesToCopy = length;
+        int srcOff = offset;
+        int j = 0;
+        while (true) {
+            BytesRef currentPage = pages.get(pageIndex + j).v();
+            int toCopyThisLoop = Math.min(pageSize - currentPageOffset, bytesToCopy);
+            System.arraycopy(b, srcOff, currentPage.bytes, currentPage.offset + currentPageOffset, toCopyThisLoop);
+            srcOff += toCopyThisLoop;
+            bytesToCopy -= toCopyThisLoop;
+            if (bytesToCopy > 0) {
+                currentPageOffset = 0;
+            } else {
+                currentPageOffset += toCopyThisLoop;
+                break;
+            }
+            j++;
+        }
+
+        // advance
+        pageIndex += j;
+    }
+
+    @Override
+    public void writeInt(int i) throws IOException {
+        if (4 > (pageSize - currentPageOffset)) {
+            super.writeInt(i);
+        } else {
+            BytesRef currentPage = pages.get(pageIndex).v();
+            VH_BE_INT.set(currentPage.bytes, currentPage.offset + currentPageOffset, i);
+            currentPageOffset += 4;
+        }
+    }
+
+    @Override
+    public void writeLong(long i) throws IOException {
+        if (8 > (pageSize - currentPageOffset)) {
+            super.writeLong(i);
+        } else {
+            BytesRef currentPage = pages.get(pageIndex).v();
+            VH_BE_LONG.set(currentPage.bytes, currentPage.offset + currentPageOffset, i);
+            currentPageOffset += 8;
+        }
+    }
+
+    @Override
+    public void reset() {
+        for (Recycler.V<BytesRef> page : pages) {
+            page.close();
```

Comment:
do we need to use IOUtils.close or something like that that keeps closing other pages if closing one of them throws an exception?